lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +212,254 @@ class Loader {
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
     const { module } = await job.run();
+
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async import(specifier, parent) {
-    const job = await this.getModuleJob(specifier, parent);
-    const { module } = await job.run();
-    return module.getNamespace();
+  async getModuleJob(specifier, parentURL) {
+    const { format, url } = await this.resolve(specifier, parentURL);
+    let job = this.moduleMap.get(url);
+    // CommonJS will set functions for lazy job evaluation.
+    if (typeof job === 'function') this.moduleMap.set(url, job = job());
+
+    if (job !== undefined) return job;
+
+    const moduleProvider = async (url, isMain) => {
+      const { format: finalFormat, source } = await this.load(url, { format });
+
+      const translator = translators.get(finalFormat);
+
+      if (!translator) throw new ERR_UNKNOWN_MODULE_FORMAT(finalFormat);
+
+      return FunctionPrototypeCall(translator, this, url, source, isMain);
+    };
+
+    const inspectBrk = (
+      parentURL === undefined &&
+      getOptionValue('--inspect-brk')
+    );
+
+    job = new ModuleJob(
+      this,
+      url,
+      moduleProvider,
+      parentURL === undefined,
+      inspectBrk
+    );
+
+    this.moduleMap.set(url, job);
+
+    return job;
   }
 
-  hook(hooks) {
-    const {
-      resolve,
-      dynamicInstantiate,
-      getFormat,
-      getSource,
-      transformSource,
-      getGlobalPreloadCode,
-    } = hooks;
+  /**
+   *
+   * @param {string | string[]} specifiers Path(s) to the module
+   * @param {string} [parentURL] Path of the parent importing the module
+   * @returns {object | object[]} A list of module export(s)
+   */
+  async import(specifiers, parentURL) {
+    const wasArr = ArrayIsArray(specifiers);
+    if (!wasArr) specifiers = [specifiers];
 
-    // Use .bind() to avoid giving access to the Loader instance when called.
-    if (resolve !== undefined)
-      this._resolve = FunctionPrototypeBind(resolve, null);
-    if (dynamicInstantiate !== undefined) {
-      process.emitWarning(
-        'The dynamicInstantiate loader hook has been removed.');
-    }
-    if (getFormat !== undefined) {
-      this._getFormat = FunctionPrototypeBind(getFormat, null);
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
-    }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
-    }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    const count = specifiers.length;
+    const namespaces = new Array(count);
+
+    for (let i = 0; i < count; i++) {
+      const specifier = specifiers[i];
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
```

Comment:
this does a linear/sequential import, can't we do this in parallel?

Reply:
Oww, actually, I think yes! The spec just says that imports are settled before anything in the module is run, but the imports sequence is indeterminate, right?

Reply:
yep, it is valid to be sequential, I am just trying to be sure we can enqueue all the module jobs eagerly rather than 1 at a time

Reply:
Hmm, doing that causes several tests to fail.

```js
    const jobs = new Array(count);

    for (let i = 0; i < count; i++) {
      jobs[i] = this.getModuleJob(specifiers[i], parentURL)
        .then((job) => job.run())
        .then(({ module }) => module.getNamespace());
    }

    const namespaces = await PromiseAll(jobs);
```

I'll investigate more on Sunday.